src/lib/scan-imports.ts

```diff
@@ -0,0 +1,217 @@
+import fs from 'fs';
+import path from 'path';
+import chalk from 'chalk';
+import ora from 'ora';
+import glob from 'glob';
+import {parse} from '@babel/parser';
+import traverse, {NodePath} from '@babel/traverse';
+import {
+  CallExpression,
+  ImportDeclaration,
+  isImportDeclaration,
+  isCallExpression,
+} from '@babel/types';
+
+interface DependencyTree {
+  [key: string]: string[];
+}
+
+const cwd = process.cwd();
+
+/**
+ * Renames `web_modules/vue.js` to `vue` based on --dest
+ */
+function getNpmName(modulePath: string, dest: string) {
+  const noExtension = modulePath.replace(/\.[^.]+$/, '');
+
+  // this RegEx allows for web_modules, web_modules/, ./folder/web_modules, etc. as valid --dest options
+  const [lastSegment] = dest.match(/\/?([^/]+)\/?$/);
+  const [_, moduleName] = noExtension.split(lastSegment);
+
+  if (!moduleName) {
+    return noExtension; // module not in --dest
+  }
+
+  const parts = moduleName.split('/').filter(part => part); // split into folders
+  const isScoped = parts[0].startsWith('@');
+  return isScoped ? `${parts[0]}/${parts[1]}` : parts[0];
+}
+
+/**
+ * Return module name from static import
+ */
+function importModuleName(node: NodePath<ImportDeclaration>) {
+  if (isImportDeclaration(node)) {
+    return node.node.source.value;
+  }
+  return undefined;
+}
+
+/**
+ * Return dynamic module name
+ */
+function dynamicImportModuleName(node: NodePath<CallExpression>) {
+  if (isCallExpression(node)) {
+    if (node.node.callee.type === 'Import' && node.node.arguments && node.node.arguments.length) {
+      return (node.node.arguments[0] as any).value;
+    }
+  }
+  return undefined;
+}
+
+/**
+ * Return array of dependencies
+ */
+function getDependencies(file: string) {
+  const code = fs.readFileSync(file, 'utf8');
+  if (!code) {
+    throw new Error(`Could not read ${file}`);
+  }
+  const ast = parse(code, {plugins: ['dynamicImport'], sourceType: 'module'});
+
+  const deps: string[] = [];
+
+  traverse(ast, {
```

Comment:
Also as we talked about, when users specify `--entry`, it shows to them how long this part takes. I found it to be only a few `ms`.

Because Babel is synchronous, there’s not an easy way to parallelize parsing. It’d probably involve Node’s [child_process](https://nodejs.org/api/child_process.html) API, but that can get messy dealing with how many max processes to spawn, keeping track of them, etc. Those can solve a problem (like in Jest’s case, dramatically increasing speed), but we should save stuff like that for when it’s needed. Overall I’m seeing this to be decently fast.

Reply:
+1, no need to optimize prematurely